File: app_manager/appmanager/src/main/java/app/editors/manager/ui/views/behaviors/CollapseToolbarEvent.java
package app.editors.manager.ui.views.behaviors;

import com.google.android.material.appbar.AppBarLayout;

public class CollapseToolbarEvent implements AppBarLayout.OnOffsetChangedListener {

    public interface OnCollapseToolbar {
        void onToolbarExpand();
        void onToolbarChange(int verticalOffset);
        void onToolbarCollapse();
    }

    private enum State {
        EXPANDED,
        COLLAPSED,
        IDLE
    }

    protected final OnCollapseToolbar mOnCollapseToolbar;
    protected State mCurrentState = State.IDLE;

    public CollapseToolbarEvent(OnCollapseToolbar onCollapseToolbar) {
        mOnCollapseToolbar = onCollapseToolbar;
    }

    @Override
    public void onOffsetChanged(AppBarLayout appBarLayout, int verticalOffset) {
        if (mOnCollapseToolbar != null) {
            if (Math.abs(verticalOffset) == appBarLayout.getTotalScrollRange()) {
                if (mCurrentState != State.COLLAPSED) {
                    mOnCollapseToolbar.onToolbarCollapse();
                    mCurrentState = State.COLLAPSED;
                }
            } else if (verticalOffset == 0) {
                if (mCurrentState != State.EXPANDED) {
                    mOnCollapseToolbar.onToolbarExpand();
                    mCurrentState = State.EXPANDED;
                }
            } else {
                if (mCurrentState != State.IDLE) {
                    mOnCollapseToolbar.onToolbarChange(verticalOffset);
                    mCurrentState = State.IDLE;
                }
            }
        }
    }

}

File: toolkit/libtoolkit/src/main/java/lib/toolkit/base/ui/views/behaviors/CollapseToolbarEvent.kt
package lib.toolkit.base.ui.views.behaviors

import com.google.android.material.appbar.AppBarLayout

class CollapseToolbarEvent(private val mOnCollapseToolbar: OnCollapseToolbar?) : AppBarLayout.OnOffsetChangedListener {

    enum class State {
        EXPANDED,
        COLLAPSED,
        IDLE
    }

    protected var mCurrentState =
        State.IDLE

    interface OnCollapseToolbar {
        fun onToolbarExpand()
        fun onToolbarChange(verticalOffset: Int)
        fun onToolbarCollapse()
    }

    override fun onOffsetChanged(appBarLayout: AppBarLayout, verticalOffset: Int) {
        mOnCollapseToolbar?.let {
            if (Math.abs(verticalOffset) == appBarLayout.totalScrollRange) {
                if (mCurrentState != State.COLLAPSED) {
                    it.onToolbarCollapse()
                    mCurrentState =
                        State.COLLAPSED
                }
            } else if (verticalOffset == 0) {
                if (mCurrentState != State.EXPANDED) {
                    it.onToolbarExpand()
                    mCurrentState =
                        State.EXPANDED
                }
            } else {
                if (mCurrentState != State.IDLE) {
                    it.onToolbarChange(verticalOffset)
                    mCurrentState =
                        State.IDLE
                }
            }
        }
    }

}
